src/main/java/cubicchunks/util/WatchersSortingList.java

```diff
@@ -0,0 +1,200 @@
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+package cubicchunks.util;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * Helper class to delay removing of an elements. Created and used to reduce CPU
+ * load during removing elements from this list on event of player movement as
+ * replacement of {@code ArrayList}.
+ */
+@SuppressWarnings({"unchecked"})
+public class WatchersSortingList<T> implements Iterable<T> {
+
+    /** Order in which list should be sorted during tick event */
+    private final Comparator<T> order;
+
+    /** Contain all data in sorting order */
+    private Object[] data = new Object[32768];
+    /**
+     * First used element of a data array. Leave some unused space to append
+     * elements with high priority.
+     */
+    private int start = data.length / 4;
+    /**
+     * Size of a list. Used to limit accessible members of data array during
+     * sorting and iteration.
+     */
+    private int size = 0;
+    /**
+     * Contain all data without certain order. Used to detect if element is
+     * already added
+     */
+    private final Set<T> dataAsSet = new HashSet<T>();
+    /** Contain elements which will be removed during tick event */
+    private final Set<T> toRemove = new HashSet<T>();
+
+    public WatchersSortingList(Comparator<T> orderIn) {
+        order = orderIn;
+    }
+
+    /** Sort elements */
+    public void sort() {
+        Arrays.sort((T[]) data, start, start + size, order);
+    }
+
+    /**
+     * Check if list is empty. Even when it return true list still could contain
+     * data, but it will be removed in attempt to access it via iterator.
+     * 
+     * @return if list contain any accessible data
+     */
+    public boolean isEmpty() {
+        return size - toRemove.size() == 0;
```

Comment:
You may want to add `assert size - toRemove.size() >= 0;` (my genIntellijRuns task enables assertions by default)